src/psyclone/psyGen.py

```diff
@@ -3273,28 +3395,22 @@ def _find_read_arguments(self, nodes):
             return []
 
         # We only need consider nodes that have arguments
-        nodes_with_args = [x for x in nodes if isinstance(x, Call) or
-                           isinstance(x, HaloExchange) or
-                           isinstance(x, GlobalSum)]
+        nodes_with_args = [x for x in nodes if
+                           isinstance(x, (Call, HaloExchange, GlobalSum))]
+        access = DataAccess(self)
         arguments = []
         for node in nodes_with_args:
             for argument in node.args:
                 # look at all arguments in our nodes
-                if argument.name != self.name:
-                    # different names so there is no dependence
-                    continue
-                if isinstance(node, HaloExchange) and \
-                   isinstance(self.call, HaloExchange):
-                    # no dependence if both nodes are halo exchanges
-                    # (as they act on different vector indices).
-                    self.call.check_vector_halos_differ(node)
-                    continue
                 if argument.access in self._read_access_types:
-                    # there is a read dependence so append to list
+                    if not access.overlaps(argument):
+                        # Accesses are independent of each other
+                        continue
                     arguments.append(argument)
                 if argument.access in self._write_access_types:
-                    # there is a write dependence so finish our search
-                    return arguments
+                    access.update_coverage(argument)
```

Comment:
Maybe it would be (slightly) nicer if `update_coverage` also returned whether or not the argument is now covered - i.e. remove the need for the subsequent `access.covered` call.